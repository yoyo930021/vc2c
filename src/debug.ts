let debugMode = false;

export function setDebugMode(value: boolean): void {
  debugMode = value;
}

export function log(message: string): void {
  if (!debugMode) return;
  console.log(`${new Date().toLocaleString()}: ${message}`);
}
