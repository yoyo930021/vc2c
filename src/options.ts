import ts from "typescript";
import { ASTConvertPlugins } from "./plugins/types";
import { getDefaultPlugins } from "./plugins";
import * as vueTemplateCompiler from "vue-template-compiler";

export interface Vc2cOptions {
  root: string;
  debug: boolean;
  compatible: boolean;
  setupPropsKey: string;
  setupContextKey: string;
  useFunctionDeclaration: boolean;
  typescript: typeof ts;
  vueTemplateCompiler: typeof vueTemplateCompiler;
  eslintConfigFile: string;
  format: boolean;
  plugins: ASTConvertPlugins;
}

export type InputVc2cOptions = Partial<Vc2cOptions>;

export function getDefaultVc2cOptions(tsModule: typeof ts = ts): Vc2cOptions {
  return {
    root: "",
    debug: false,
    compatible: false,
    setupPropsKey: "props",
    setupContextKey: "context",
    useFunctionDeclaration: false,
    typescript: tsModule,
    vueTemplateCompiler: vueTemplateCompiler,
    eslintConfigFile: "eslintrc.config.mjs",
    format: false,
    plugins: getDefaultPlugins(tsModule),
  };
}

export function mergeVc2cOptions(
  original: Vc2cOptions,
  merged: InputVc2cOptions,
): Vc2cOptions {
  return {
    ...original,
    ...merged,
  };
}
