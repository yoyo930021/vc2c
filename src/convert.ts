import type ts from "typescript";
import { getDecoratorNames, getDefaultExportNode } from "./utils";
import { runPlugins } from "./plugins";
import { Vc2cOptions } from "./options";
import { log } from "./debug";

const vueClassModules = [
  "vue-class-component",
  "vue-property-decorator",
  "nuxt-property-decorator",
];

export function convertAST(
  sourceFile: ts.SourceFile,
  options: Vc2cOptions,
  program: ts.Program,
): string {
  const tsModule = options.typescript;

  log("check vue class library");
  const vueClassModuleImportStatement = sourceFile.statements.find(
    (statement) => {
      if (tsModule.isImportDeclaration(statement)) {
        if (
          vueClassModules.includes(
            (statement.moduleSpecifier as ts.StringLiteral).text,
          )
        ) {
          return true;
        }
      }
      return false;
    },
  );
  if (!vueClassModuleImportStatement) {
    throw new Error("no vue class library in this file.");
  }

  log("check default export class");
  const defaultExportNode = getDefaultExportNode(
    options.typescript,
    sourceFile,
  );
  if (!defaultExportNode) {
    throw new Error("no default export class");
  }

  const otherStatements = sourceFile.statements
    .map((el) => el)
    .filter(
      (el) =>
        !(
          (tsModule.isClassDeclaration(el) &&
            getDecoratorNames(tsModule, el).includes("Component")) ||
          (tsModule.isImportDeclaration(el) &&
            vueClassModules.includes(
              (el.moduleSpecifier as ts.StringLiteral).text,
            )) ||
          (tsModule.isImportDeclaration(el) &&
            (el.moduleSpecifier as ts.StringLiteral).text === "vue")
        ),
    );

  const resultStatements = [
    ...otherStatements,
    ...runPlugins(defaultExportNode, options, program),
  ];

  log("output result code");
  const printer = tsModule.createPrinter();

  return printer.printFile(
    tsModule.factory.updateSourceFile(sourceFile, resultStatements),
  );
}
