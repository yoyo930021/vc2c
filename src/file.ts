import { isVueFile, parseVueFile } from "./utils";
import { Vc2cOptions } from "./options";
import { log } from "./debug";

export enum FileKind {
  VUE,
  TS,
}

export type FileInfo =
  | {
      fsPath: string;
      kind: FileKind;
      content: string;
    }
  | {
      fsPath: string;
      kind: FileKind;
      fileContent: string;
      start: number;
      end: number;
      content: string;
    };

export function parseVueSFCOrTsFile(
  filePath: string,
  fileContent: string,
  options: Vc2cOptions,
): FileInfo {
  if (isVueFile(filePath)) {
    const scriptContent = parseVueFile(
      options.vueTemplateCompiler,
      fileContent,
    ).script;
    if (scriptContent) {
      log(`Readed Vue file: ${filePath}`);
      return {
        fsPath: filePath,
        kind: FileKind.VUE,
        start: scriptContent.start,
        end: scriptContent.end,
        content: scriptContent.content,
        fileContent,
      };
    }
    throw new Error("The Vue SFC don't have sciprt element.");
  } else {
    log(`Readed TS file: ${filePath}`);
    return {
      fsPath: filePath,
      kind: FileKind.TS,
      content: fileContent,
    };
  }
}
