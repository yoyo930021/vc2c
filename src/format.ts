import { Vc2cOptions } from "./options";
import path from "node:path";
import fs from "node:fs";
import { log } from "./debug";
import prettier from "prettier/standalone";
import tsParser from "prettier/plugins/typescript";
import estree from "prettier/plugins/estree";

/* v8 ignore start */
export async function format(content: string, options: Vc2cOptions) {
  const isNode = typeof window === "undefined";
  if (!isNode) {
    return prettier.format(content, {
      parser: "typescript",
      plugins: [tsParser, estree],
      semi: true,
      singleQuote: true,
    });
  }

  const eslintConfigPath = path.resolve(options.root, options.eslintConfigFile);
  const prettierFormat = await import("prettier-eslint").then((m) => m.default);
  const prettierEslintOptions = fs.existsSync(eslintConfigPath)
    ? {
        text: content,
        filePath: eslintConfigPath,
        prettierOptions: {
          parser: "typescript",
        },
        fallbackPrettierOptions: {
          parser: "typescript",
        },
      }
    : {
        text: content,
        eslintConfig: {
          parser: require.resolve("@typescript-eslint/parser"),
          parserOptions: {
            sourceType: "module",
            ecmaFeatures: {
              jsx: false,
            },
          },
          rules: {
            semi: ["error", "never"],
            "padding-line-between-statements": [
              "error",
              { blankLine: "always", prev: "*", next: "export" },
              { blankLine: "always", prev: "const", next: "*" },
              { blankLine: "always", prev: "*", next: "const" },
            ],
          },
        },
        prettierOptions: {
          parser: "typescript",
          Semicolons: false,
          singleQuote: true,
        },
        fallbackPrettierOptions: {
          parser: "typescript",
          singleQuote: true,
          Semicolons: false,
        },
      };

  log("Format result code.....");
  return prettierFormat(prettierEslintOptions);
}
/* v8 ignore stop */
