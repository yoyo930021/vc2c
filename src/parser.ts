import ts from 'typescript'
import { Vc2cOptions } from './options'

export const defaultCompilerOptions: ts.CompilerOptions = {
  allowNonTsExtensions: true,
  allowJs: true,
  checkJs: true,
  noEmit: true
}

export function getSingleFileProgram (content: string, options: Vc2cOptions) {
  const fileName = 'ast.ts'
  const tsModule = options.typescript

  const compilerHost: ts.CompilerHost = {
    fileExists (filePath: string) {
      return filePath.includes(fileName)
    },
    getCanonicalFileName () {
      return fileName
    },
    getCurrentDirectory () {
      return ''
    },
    getDirectories () {
      return []
    },
    getDefaultLibFileName () {
      return 'lib.d.ts'
    },
    getNewLine () {
      return '\n'
    },
    getSourceFile (filename: string) {
      return tsModule.createSourceFile(filename, content, ts.ScriptTarget.Latest, true)
    },
    readFile () {
      return undefined
    },
    useCaseSensitiveFileNames () {
      return true
    },
    writeFile () {
      return null
    }
  }

  const program = tsModule.createProgram(
    [fileName],
    {
      noResolve: true,
      target: tsModule.ScriptTarget.Latest,
      ...defaultCompilerOptions
    },
    compilerHost
  )

  const ast = program.getSourceFile(fileName)
  if (!ast) {
    throw new Error('Can\'t convert code to TypeScript AST.')
  }

  return { ast, program }
}
