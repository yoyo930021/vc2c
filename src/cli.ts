import program from 'commander'
import inquirer from 'inquirer'
import { version } from '../package.json'
import { writeFileInfo } from './file'

const { convertFile } = require('./index.js')

function camelize (str: string) {
  return str.replace(/-(\w)/g, (_, c) => c ? c.toUpperCase() : '')
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getCmdOptions (cmd: any) {
  const args: { [key: string]: boolean | string } = {}
  cmd.options.forEach((o: { long: string }) => {
    const key = camelize(o.long.replace(/^--/, ''))

    if (typeof cmd[key] !== 'function' && typeof cmd[key] !== 'undefined') {
      args[key] = cmd[key]
    }
  })
  return args
}

program
  .version(version)
  .usage('<command> [options]')

program
  .command('single <filePath>')
  .description('convert vue component file from class to composition api')
  .option('-v, --view', 'Output file content on stdout, and no write file.')
  .option('-o, --output', 'Output result file path.')
  .option('-r, --root <root>', 'Set root path for calc file absolute path. Default:`process.cwd()`')
  .option('-c, --config <config>', 'Set vc2c config file path. Default: `\'.vc2c.js\'`')
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  .action(async (filePath: string, cmd) => {
    const cmdOptions = getCmdOptions(cmd)
    if (!cmdOptions.output && !cmdOptions.view) {
      const result = await inquirer.prompt({
        name: 'ok',
        type: 'confirm',
        message: 'You aren\'t using -o option to set output file path, It will replace original file content.'
      })
      if (!result.ok) {
        return
      }
    }

    const { file, result } = convertFile(filePath, cmdOptions.root as string, cmdOptions.config as string)
    if (cmdOptions.view) {
      console.log(result)
      return
    }

    writeFileInfo(file, result)
    console.log('Please check the TODO comments on result.')
  })

program.parse(process.argv)
