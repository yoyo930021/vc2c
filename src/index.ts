import { getSingleFileProgram } from "./parser";
import { convertAST } from "./convert";
import {
  InputVc2cOptions,
  getDefaultVc2cOptions,
  mergeVc2cOptions,
} from "./options";
import { format } from "./format";
import { parseVueSFCOrTsFile } from "./file";
import { setDebugMode } from "./debug";
import * as BuiltInPlugins from "./plugins/builtIn";

export async function convert(
  filePath: string,
  fileContent: string,
  inputOptions: InputVc2cOptions = {},
): Promise<string> {
  const options = mergeVc2cOptions(
    getDefaultVc2cOptions(inputOptions.typescript),
    inputOptions,
  );

  if (options.debug) {
    setDebugMode(true);
  }

  const { content } = parseVueSFCOrTsFile(filePath, fileContent, options);
  const { ast, program } = getSingleFileProgram(content, options);
  const result = convertAST(ast, options, program);

  return options.format ? format(result, options) : result;
}

export * from "./plugins/types";
export { BuiltInPlugins };
export * from "./utils";
export { getDefaultVc2cOptions, type Vc2cOptions } from "./options";
