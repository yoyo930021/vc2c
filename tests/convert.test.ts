import { describe, expect, test } from "vitest";
import { readFileSync } from "node:fs";
import { convert } from "../src";

describe("converter", () => {
  test("covert component name", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/ComponentName.vue").toString(),
      ),
    ).toMatchSnapshot();
  });

  test("covert data", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Data.vue").toString(),
      ),
    ).toMatchSnapshot();
  });

  test("covert props", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Props.vue").toString(),
      ),
    ).toMatchSnapshot();
  });

  test("covert props array", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/PropsArray.vue").toString(),
      ),
    ).toMatchSnapshot();
  });

  test("covert methods", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Methods.vue").toString(),
      ),
    ).toMatchSnapshot();
  });

  test("covert methods with the function declaration", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Methods.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert computed", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Computed.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert watch", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Watch.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert lifecycle hooks", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/IntervalHook.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert render function", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Render.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert provide", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Provide.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert inject", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Inject.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert emit", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Emit.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert dom ref", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/DomRef.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });

  test("covert model", async () => {
    expect(
      await convert(
        "index.vue",
        readFileSync(__dirname + "/fixture/Model.vue").toString(),
        {
          useFunctionDeclaration: true,
        },
      ),
    ).toMatchSnapshot();
  });
});
