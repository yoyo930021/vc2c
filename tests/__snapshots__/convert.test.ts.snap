// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`converter > covert component name 1`] = `
"import { type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "IndexPage",
    setup(props, context) {
        // TODO: Please remove unused return variable
        return {};
    }
});
"
`;

exports[`converter > covert computed 1`] = `
"interface Item {
    id: number;
    name: string;
    active: boolean;
}
import { computed, type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {
        items: { type: [Object, Array] as PropType<Item[] | Item>, default: null },
        firstName: { type: String, default: "" },
        lastName: { type: String, default: "" }
    },
    name: "MyComponent",
    setup(props, context) {
        const count = computed(() => {
            return /* TODO: Check this convert result, it can work well in 80% case.*/ context.root.$store.state.count;
        });
        function getFullName(firstName: string, lastName: string) {
            return \`\${firstName} \${lastName}\`.trim();
        }
        const activeItems = computed(() => {
            return props.items.filter((item) => item.selected);
        });
        const fullName = computed({
            get: () => {
                return getFullName(props.firstName, props.lastName);
            },
            set: (v: string) => {
                const [firstName = "", lastName = ""] = v.split(" ");
                context.emit("update:firstName", firstName);
                context.emit("update:lastName", lastName);
            }
        });
        // TODO: Please remove unused return variable
        return { count, getFullName, activeItems, fullName };
    }
});
"
`;

exports[`converter > covert data 1`] = `
"import { type PropType, reactive, toRefs, ref, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "MyComponent",
    setup(props, context) {
        const { isLoading, loginForm } = (() => {
            return toRefs(reactive({
                isLoading: false,
                loginForm: {
                    email: "",
                    password: "",
                },
            }));
        })();
        const msg = ref("");
        const name = ref<string | null>(null);
        const user = ref({
            first_name: "John",
            last_name: "Doe" as string | null,
            age: 42,
        });
        // TODO: Please remove unused return variable
        return { isLoading, loginForm, msg, name, user };
    }
});
"
`;

exports[`converter > covert dom ref 1`] = `
"import AnotherComponent from "@/path/to/another-component.vue";
import { type PropType, ref, defineComponent } from "vue";
export default defineComponent({
    props: {},
    name: "MyComponent",
    setup(props, context) {
        const anotherComponent = ref<AnotherComponent>(null);
        const button = ref<HTMLButtonElement>(null);
        // TODO: Please remove unused return variable
        return { anotherComponent, button };
    }
});
"
`;

exports[`converter > covert emit 1`] = `
"interface Item {
    id: number;
    name: string;
}
import { type PropType, ref, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "MyComponent",
    setup(props, context) {
        const currentValue = ref("");
        function onChange() {
            context.emit("on-change", currentValue.value);
        }
        function onInput() {
            context.emit("update:model-value", currentValue.value);
        }
        function reset() {
            currentValue.value = "";
            context.emit("reset");
        }
        // TODO: Please remove unused return variable
        return { currentValue, onChange, onInput, reset };
    }
});
"
`;

exports[`converter > covert inject 1`] = `
"const symbol = Symbol("baz");
import { type PropType, inject, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "MyComponent",
    setup(props, context) {
        const foo = inject<string>("foo");
        const bar = inject<string>("bar");
        const optional = inject<string>("optional", "default");
        const baz = inject<string>(symbol);
        // TODO: Please remove unused return variable
        return { foo, bar, optional, baz };
    }
});
"
`;

exports[`converter > covert lifecycle hooks 1`] = `
"import { type PropType, onMounted, onBeforeUpdate, onUpdated, onBeforeUnmount, onUnmounted, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "MyComponent",
    setup(props, context) {
        console.log("before create");
        console.log("created");
        onMounted(() => {
            console.log("mounted");
        });
        onBeforeUpdate(() => {
            console.log("before update");
        });
        onUpdated(() => {
            console.log("updated");
        });
        onBeforeUnmount(() => {
            console.log("before destroy");
        });
        onUnmounted(() => {
            console.log("destroyed");
        });
        // TODO: Please remove unused return variable
        return { onMounted, onBeforeUpdate, onUpdated, onBeforeUnmount, onUnmounted };
    }
});
"
`;

exports[`converter > covert methods 1`] = `
"interface Item {
    id: number;
    name: string;
}
import { type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {
        userId: {
            type: Number,
            required: true,
        },
        userName: { type: String as PropType<string | null>, default: null }
    },
    name: "MyComponent",
    setup(props, context) {
        /**
         * Returns {void}
         */
        const noop = (): void => { };
        const greetings = () => {
            console.log(\`hello \${props.userName ?? "Unknown"} (id: \${props.userId})\`);
        };
        const fn = () => {
            noop();
            greetings();
        };
        // TODO: Please remove unused return variable
        return { noop, greetings, fn };
    }
});
"
`;

exports[`converter > covert methods with the function declaration 1`] = `
"interface Item {
    id: number;
    name: string;
}
import { type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {
        userId: {
            type: Number,
            required: true,
        },
        userName: { type: String as PropType<string | null>, default: null }
    },
    name: "MyComponent",
    setup(props, context) {
        /**
         * Returns {void}
         */
        function noop(): void { }
        function greetings() {
            console.log(\`hello \${props.userName ?? "Unknown"} (id: \${props.userId})\`);
        }
        function fn() {
            noop();
            greetings();
        }
        // TODO: Please remove unused return variable
        return { noop, greetings, fn };
    }
});
"
`;

exports[`converter > covert model 1`] = `
"import { type PropType, defineComponent } from "vue";
export default defineComponent({
    model: {
        prop: "checked",
        event: "change"
    },
    props: {
        checked: { type: Boolean }
    },
    name: "YourComponent",
    setup(props, context) {
        // TODO: Please remove unused return variable
        return {};
    }
});
"
`;

exports[`converter > covert props 1`] = `
"interface Item {
    id: number;
    name: string;
}
import { type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {
        msg: null,
        name: {},
        loading: Boolean,
        focused: {
            type: Boolean,
            default: true,
        },
        disabled: {
            type: Boolean,
        },
        propName: { type: String as PropType<string | null>, default: null },
        selected: { type: [Object, Array] as PropType<Item[] | Item>, default: null }
    },
    name: "MyComponent",
    setup(props, context) {
        // TODO: Please remove unused return variable
        return {};
    }
});
"
`;

exports[`converter > covert props array 1`] = `
"interface Item {
    id: number;
    name: string;
}
import { type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {
        msg: null,
        name: null,
        loading: null,
        focused: null,
        disabled: null,
        propName: { type: String as PropType<string | null>, default: null },
        selected: { type: [Object, Array] as PropType<Item[] | Item>, default: null }
    },
    name: "MyComponent",
    setup(props, context) {
        // TODO: Please remove unused return variable
        return {};
    }
});
"
`;

exports[`converter > covert provide 1`] = `
"const symbol = Symbol("baz");
import { type PropType, provide, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "MyComponent",
    setup(props, context) {
        provide("foo", "foo");
        provide("bar", "bar");
        provide(symbol, "fiz");
        // TODO: Please remove unused return variable
        return {};
    }
});
"
`;

exports[`converter > covert render function 1`] = `
"import { type PropType, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {},
    name: "MyComponent",
    render(h) {
        return h("div");
    },
    setup(props, context) {
        // TODO: Please remove unused return variable
        return {};
    }
});
"
`;

exports[`converter > covert watch 1`] = `
"import { type PropType, computed, watch, defineComponent } from "vue";
/**
 * My basic tag
 */
export default defineComponent({
    props: {
        firstName: { type: String, default: "" },
        lastName: { type: String, default: "" }
    },
    name: "MyComponent",
    setup(props, context) {
        function getFullName(firstName: string, lastName: string) {
            return \`\${firstName} \${lastName}\`.trim();
        }
        const fullName = computed(() => {
            return getFullName(props.firstName, props.lastName);
        });
        watch(firstName, onFirstNameChanged);
        function onFirstNameChanged() {
            console.log("changed first name", props.firstName);
        }
        watch(fullName, greetings, { immediate: true });
        function greetings() {
            console.log("hello", fullName.value);
        }
        // TODO: Please remove unused return variable
        return { getFullName, fullName };
    }
});
"
`;
